src/osmdata.cpp

```diff
@@ -389,16 +393,39 @@ progress_display_t osmdata_t::process_file(osmium::io::File const &file,
     return handler.progress();
 }
 
-void osmdata_t::process_stage1b() const
+void osmdata_t::process_stage1bc() const
 {
-    if (m_dependency_manager->has_pending()) {
-        multithreaded_processor proc{m_conninfo, m_mid, m_outs,
-                                     (std::size_t)m_num_procs};
+    if (!m_dependency_manager->has_pending()) {
+        return;
+    }
+
+    multithreaded_processor proc{m_conninfo, m_mid, m_outs,
+                                 (std::size_t)m_num_procs};
+
+    // stage 1b processing
+    proc.process_ways(m_dependency_manager->get_pending_way_ids());
+    proc.process_relations(m_dependency_manager->get_pending_relation_ids());
+    proc.merge_expire_trees();
+
+    // If none of the outputs wants to do stage 2 processing, we can skip
+    // stage 1c.
+    if (std::none_of(m_outs.cbegin(), m_outs.cend(),
+                     [](std::shared_ptr<output_t> const &out) {
+                         return out->has_stage2_processing();
+                     })) {
+        return;
+    }
```

Comment:
If there is no stage2 processing, get_marked_way_ids()  simply is empty. So the loop below seems enough.

Reply:
Okay. Fixed.